File: packages/mongodb/src/driver.ts
import { Buffer } from "node:buffer";
import { Binary, MongoClient, type Collection as MongoCollection } from "mongodb";

import {
  type Batch,
  type BatchResult,
  type Capabilities,
  type CollectionName,
  type DeleteOptions,
  type Driver,
  type Key,
  type Namespace,
  type ReadOptions,
  type ScanOptions,
  type StoredValue,
  type WriteOptions,
  YugenDbError,
  isExpired,
  minimalCapabilities,
  normaliseKey,
} from "@yugendb/core";

import { mongodbDocumentSchema, storageCollectionName, type MongoDbDocumentSchema } from "./schema.js";

/** Options used to connect the MongoDB driver. */
export interface MongoDbDriverOptions {
  readonly connectionString: string;
  readonly database?: string;
  readonly storageCollection?: string;
  readonly createIndexesOnInitialise?: boolean;
}

interface StoredDocument {
  namespace: string;
  collection: string;
  key: string;
  value: Binary | Uint8Array;
  codec: string;
  createdAt: string;
  updatedAt: string;
  expiresAt: string | null;
}

/** MongoDB driver for yugendb document-backed storage. */
export class MongoDbDriver implements Driver {
  readonly name = "mongodb";
  readonly options: Required<MongoDbDriverOptions>;
  private client: MongoClient | null = null;

  constructor(options: MongoDbDriverOptions | string) {
    this.options = typeof options === "string"
      ? {
          connectionString: options,
          database: "yugendb",
          storageCollection: storageCollectionName,
          createIndexesOnInitialise: true,
        }
      : {
          connectionString: options.connectionString,
          database: options.database ?? "yugendb",
          storageCollection: options.storageCollection ?? storageCollectionName,
          createIndexesOnInitialise: options.createIndexesOnInitialise ?? true,
        };
  }

  capabilities(): Capabilities {
    return mongodbCapabilities();
  }

  schema(): MongoDbDocumentSchema {
    return mongodbDocumentSchema();
  }

  async initialise(): Promise<void> {
    if (this.client !== null) {
      return;
    }

    this.client = new MongoClient(this.options.connectionString);
    await this.client.connect();

    if (this.options.createIndexesOnInitialise) {
      const collection = this.documents();
      await collection.createIndex(
        { namespace: 1, collection: 1, key: 1 },
        { unique: true, name: "idx_yugendb_identity" },
      );
      await collection.createIndex({ expiresAt: 1 }, { name: "idx_yugendb_expires_at" });
    }
  }

  async finalise(): Promise<void> {
    await this.client?.close();
    this.client = null;
  }

  async get(
    namespace: Namespace,
    collection: CollectionName,
    key: Key,
    options: ReadOptions,
  ): Promise<StoredValue | null> {
    const document = await this.documents().findOne({ namespace, collection, key });
    if (document === null) {
      return null;
    }

    const value = documentToStoredValue(document);
    return isExpired(value) && options.allowExpired !== true ? null : value;
  }

  async set(
    namespace: Namespace,
    collection: CollectionName,
    key: Key,
    value: StoredValue,
    options: WriteOptions,
  ): Promise<void> {
    if (options.overwrite === false && await this.exists(namespace, collection, key)) {
      throw new YugenDbError({
        code: "CONFLICT",
        message: "MongoDB driver refused to overwrite an existing value.",
      });
    }

    await this.documents().updateOne(
      { namespace, collection, key },
      { $set: storedValueToDocument(namespace, collection, key, value) },
      { upsert: true },
    );
  }

  async delete(
    namespace: Namespace,
    collection: CollectionName,
    key: Key,
    options: DeleteOptions,
  ): Promise<boolean> {
    const result = await this.documents().deleteOne({ namespace, collection, key });
    const removed = result.deletedCount > 0;

    if (options.mustExist === true && !removed) {
      throw new YugenDbError({
        code: "NOT_FOUND",
        message: "MongoDB driver could not delete a missing value.",
      });
    }

    return removed;
  }

  async exists(namespace: Namespace, collection: CollectionName, key: Key): Promise<boolean> {
    return (await this.get(namespace, collection, key, { allowExpired: false })) !== null;
  }

  async scanPrefix(
    namespace: Namespace,
    collection: CollectionName,
    options: ScanOptions,
  ): Promise<Array<[Key, StoredValue]>> {
    const prefix = options.prefix ?? "";
    const cursor = this.documents()
      .find({ namespace, collection, key: { $regex: `^${escapeRegExp(prefix)}` } })
      .sort({ key: 1 });
    const values: Array<[Key, StoredValue]> = [];

    for await (const document of cursor) {
      const value = documentToStoredValue(document);
      if (isExpired(value) && options.includeExpired !== true) {
        continue;
      }
      values.push([normaliseKey(document.key), value]);
      if (options.limit !== undefined && values.length >= options.limit) {
        break;
      }
    }

    return values;
  }

  async batch(batch: Batch): Promise<BatchResult> {
    let appliedOperations = 0;
    let deletedKeys = 0;

    for (const operation of batch.operations) {
      if (operation.kind === "set") {
        await this.set(operation.namespace, operation.collection, operation.key, operation.value, operation.options ?? {});
        appliedOperations += 1;
      } else {
        if (await this.delete(operation.namespace, operation.collection, operation.key, operation.options ?? {})) {
          deletedKeys += 1;
        }
        appliedOperations += 1;
      }
    }

    return { appliedOperations, deletedKeys };
  }

  private documents(): MongoCollection<StoredDocument> {
    if (this.client === null) {
      throw new YugenDbError({
        code: "CONNECTION_ERROR",
        message: "MongoDB driver has not been initialised.",
      });
    }
    return this.client.db(this.options.database).collection<StoredDocument>(this.options.storageCollection);
  }
}

/** Returns the capabilities reported by the MongoDB driver. */
export function mongodbCapabilities(): Capabilities {
  return {
    ...minimalCapabilities(),
    ttl: true,
    prefixScan: true,
    batchWrite: true,
    connectionPooling: true,
  };
}

/** Creates a MongoDB yugendb driver. */
export function mongodb(
  connectionString: string,
  options: Omit<MongoDbDriverOptions, "connectionString"> = {},
): MongoDbDriver {
  return new MongoDbDriver({ connectionString, ...options });
}

function storedValueToDocument(
  namespace: Namespace,
  collection: CollectionName,
  key: Key,
  value: StoredValue,
): StoredDocument {
  return {
    namespace,
    collection,
    key,
    value: new Binary(Buffer.from(value.bytes)),
    codec: value.metadata.codec,
    createdAt: value.metadata.createdAt.toISOString(),
    updatedAt: value.metadata.updatedAt.toISOString(),
    expiresAt: value.metadata.expiresAt instanceof Date ? value.metadata.expiresAt.toISOString() : null,
  };
}

function documentToStoredValue(document: StoredDocument): StoredValue {
  const raw = document.value instanceof Binary ? document.value.buffer : document.value;
  return {
    bytes: new Uint8Array(raw),
    metadata: {
      codec: document.codec,
      createdAt: new Date(document.createdAt),
      updatedAt: new Date(document.updatedAt),
      expiresAt: document.expiresAt === null ? null : new Date(document.expiresAt),
    },
  };
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/gu, "\\$&");
}

File: crates/yugendb-mongodb/src/driver.rs
//! MongoDB driver implementation.
//!
//! MongoDB stores yugendb records as documents. MongoDB remains an
//! implementation detail. Applications use the same namespace, collection, key,
//! and serialised value API as the other yugendb drivers.

use async_trait::async_trait;
use chrono::{DateTime, Utc};
use mongodb::{
    bson::{doc, spec::BinarySubtype, Binary, Bson, Document},
    options::{FindOptions, IndexOptions, UpdateOptions},
    Client, Collection, IndexModel,
};
use parking_lot::Mutex;
use yugendb_core::{
    Batch, BatchOperation, BatchResult, Capabilities, CollectionName, DeleteOptions, Driver, Key,
    Namespace, ReadOptions, Result, ScanOptions, StoredValue, ValueBytes, ValueMetadata,
    WriteOptions, YugenDbError,
};

use crate::document::{
    CODEC_FIELD, COLLECTION_FIELD, CREATED_AT_FIELD, EXPIRES_AT_FIELD, KEY_FIELD, NAMESPACE_FIELD,
    STORAGE_COLLECTION_NAME, UPDATED_AT_FIELD, VALUE_FIELD,
};

/// Options used to configure the MongoDB driver.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MongoDbDriverOptions {
    /// MongoDB connection string.
    pub connection_string: String,
    /// Database name to store yugendb documents in.
    pub database: String,
    /// Backing MongoDB collection name.
    pub storage_collection: String,
    /// Whether indexes should be created during initialisation.
    pub create_indexes_on_initialise: bool,
}

impl MongoDbDriverOptions {
    /// Creates default MongoDB options for a connection string.
    #[must_use]
    pub fn new(connection_string: impl Into<String>) -> Self {
        Self {
            connection_string: connection_string.into(),
            database: "yugendb".to_string(),
            storage_collection: STORAGE_COLLECTION_NAME.to_string(),
            create_indexes_on_initialise: true,
        }
    }
}

/// MongoDB document-backed driver.
#[derive(Debug)]
pub struct MongoDbDriver {
    options: MongoDbDriverOptions,
    client: Mutex<Option<Client>>,
}

impl MongoDbDriver {
    /// Creates a new MongoDB driver using default options.
    #[must_use]
    pub fn new(connection_string: impl Into<String>) -> Self {
        Self::with_options(MongoDbDriverOptions::new(connection_string))
    }

    /// Creates a new MongoDB driver with explicit options.
    #[must_use]
    pub fn with_options(options: MongoDbDriverOptions) -> Self {
        Self {
            options,
            client: Mutex::new(None),
        }
    }

    /// Returns configured options.
    #[must_use]
    pub const fn options(&self) -> &MongoDbDriverOptions {
        &self.options
    }

    fn map_error(error: mongodb::error::Error) -> YugenDbError {
        YugenDbError::driver_error("mongodb", error.to_string())
    }

    fn client(&self) -> Result<Client> {
        self.client.lock().as_ref().cloned().ok_or_else(|| {
            YugenDbError::connection_error("MongoDB driver has not been initialised")
        })
    }

    fn store_collection(&self) -> Result<Collection<Document>> {
        let client = self.client()?;
        Ok(client
            .database(&self.options.database)
            .collection(&self.options.storage_collection))
    }

    fn identity_filter(namespace: &Namespace, collection: &CollectionName, key: &Key) -> Document {
        let mut filter = Document::new();
        filter.insert(NAMESPACE_FIELD, namespace.as_str());
        filter.insert(COLLECTION_FIELD, collection.as_str());
        filter.insert(KEY_FIELD, key.as_str());
        filter
    }

    fn document_from_value(
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        value: &StoredValue,
    ) -> Document {
        let mut document = Document::new();
        document.insert(NAMESPACE_FIELD, namespace.as_str());
        document.insert(COLLECTION_FIELD, collection.as_str());
        document.insert(KEY_FIELD, key.as_str());
        document.insert(
            VALUE_FIELD,
            Bson::Binary(Binary {
                subtype: BinarySubtype::Generic,
                bytes: value.bytes.as_slice().to_vec(),
            }),
        );
        document.insert(CODEC_FIELD, value.metadata.codec.clone());
        document.insert(CREATED_AT_FIELD, value.metadata.created_at.to_rfc3339());
        document.insert(UPDATED_AT_FIELD, value.metadata.updated_at.to_rfc3339());
        match value.metadata.expires_at {
            Some(expires_at) => document.insert(EXPIRES_AT_FIELD, expires_at.to_rfc3339()),
            None => document.insert(EXPIRES_AT_FIELD, Bson::Null),
        };
        document
    }

    fn value_from_document(document: Document) -> Result<StoredValue> {
        let bytes = match document.get(VALUE_FIELD) {
            Some(Bson::Binary(binary)) => binary.bytes.clone(),
            _ => {
                return Err(YugenDbError::driver_error(
                    "mongodb",
                    "stored document has invalid value field",
                ))
            }
        };
        let codec = document
            .get_str(CODEC_FIELD)
            .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?
            .to_owned();
        let created_at = DateTime::parse_from_rfc3339(
            document
                .get_str(CREATED_AT_FIELD)
                .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?,
        )
        .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?
        .with_timezone(&Utc);
        let updated_at = DateTime::parse_from_rfc3339(
            document
                .get_str(UPDATED_AT_FIELD)
                .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?,
        )
        .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?
        .with_timezone(&Utc);
        let expires_at = match document.get(EXPIRES_AT_FIELD) {
            Some(Bson::String(value)) => Some(
                DateTime::parse_from_rfc3339(value)
                    .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?
                    .with_timezone(&Utc),
            ),
            Some(Bson::Null) | None => None,
            _ => {
                return Err(YugenDbError::driver_error(
                    "mongodb",
                    "stored document has invalid expiresAt field",
                ))
            }
        };
        Ok(StoredValue::new(
            ValueBytes::from(bytes),
            ValueMetadata {
                codec,
                created_at,
                updated_at,
                expires_at,
            },
        ))
    }
}

impl Clone for MongoDbDriver {
    fn clone(&self) -> Self {
        Self::with_options(self.options.clone())
    }
}

/// Capabilities for the implemented MongoDB driver.
#[must_use]
pub const fn mongodb_capabilities() -> Capabilities {
    Capabilities {
        transactions: false,
        ttl: true,
        prefix_scan: true,
        atomic_increment: false,
        batch_write: true,
        raw_sql: false,
        document_query: false,
        json_query: false,
        migrations: false,
        connection_pooling: true,
        watch: false,
        backup: false,
    }
}

/// Convenience constructor for the MongoDB driver.
#[must_use]
pub fn mongodb(connection_string: impl Into<String>) -> MongoDbDriver {
    MongoDbDriver::new(connection_string)
}

#[async_trait]
impl Driver for MongoDbDriver {
    fn name(&self) -> &'static str {
        "mongodb"
    }
    fn capabilities(&self) -> Capabilities {
        mongodb_capabilities()
    }

    async fn initialise(&self) -> Result<()> {
        let client = Client::with_uri_str(&self.options.connection_string)
            .await
            .map_err(Self::map_error)?;
        if self.options.create_indexes_on_initialise {
            let collection = client
                .database(&self.options.database)
                .collection::<Document>(&self.options.storage_collection);
            let identity = IndexModel::builder()
                .keys({
                    let mut keys = Document::new();
                    keys.insert(NAMESPACE_FIELD, 1);
                    keys.insert(COLLECTION_FIELD, 1);
                    keys.insert(KEY_FIELD, 1);
                    keys
                })
                .options(
                    IndexOptions::builder()
                        .unique(true)
                        .name("idx_yugendb_identity".to_owned())
                        .build(),
                )
                .build();
            collection
                .create_index(identity, None)
                .await
                .map_err(Self::map_error)?;
            collection
                .create_index(
                    IndexModel::builder()
                        .keys({
                            let mut keys = Document::new();
                            keys.insert(EXPIRES_AT_FIELD, 1);
                            keys
                        })
                        .build(),
                    None,
                )
                .await
                .map_err(Self::map_error)?;
        }
        *self.client.lock() = Some(client);
        Ok(())
    }

    async fn finalise(&self) -> Result<()> {
        *self.client.lock() = None;
        Ok(())
    }

    async fn get(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        options: &ReadOptions,
    ) -> Result<Option<StoredValue>> {
        let store = self.store_collection()?;
        let document = store
            .find_one(Self::identity_filter(namespace, collection, key), None)
            .await
            .map_err(Self::map_error)?;
        let Some(document) = document else {
            return Ok(None);
        };
        let value = Self::value_from_document(document)?;
        Ok((options.allow_expired || !value.is_expired()).then_some(value))
    }

    async fn set(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        value: StoredValue,
        options: &WriteOptions,
    ) -> Result<()> {
        if !options.overwrite && self.exists(namespace, collection, key).await? {
            return Err(YugenDbError::conflict(
                "MongoDB driver refused to overwrite an existing value",
            ));
        }
        let store = self.store_collection()?;
        let document = Self::document_from_value(namespace, collection, key, &value);
        store
            .update_one(
                Self::identity_filter(namespace, collection, key),
                doc! { "$set": document },
                Some(UpdateOptions::builder().upsert(true).build()),
            )
            .await
            .map_err(Self::map_error)?;
        Ok(())
    }

    async fn delete(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
        options: &DeleteOptions,
    ) -> Result<bool> {
        let store = self.store_collection()?;
        let result = store
            .delete_one(Self::identity_filter(namespace, collection, key), None)
            .await
            .map_err(Self::map_error)?;
        let removed = result.deleted_count > 0;
        if options.must_exist && !removed {
            return Err(YugenDbError::not_found(
                "MongoDB driver could not delete a missing value",
            ));
        }
        Ok(removed)
    }

    async fn exists(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        key: &Key,
    ) -> Result<bool> {
        Ok(self
            .get(namespace, collection, key, &ReadOptions::default())
            .await?
            .is_some())
    }

    async fn scan_prefix(
        &self,
        namespace: &Namespace,
        collection: &CollectionName,
        options: &ScanOptions,
    ) -> Result<Vec<(Key, StoredValue)>> {
        let store = self.store_collection()?;
        let prefix = regex_escape(options.prefix.as_deref().unwrap_or(""));
        let mut key_filter = Document::new();
        key_filter.insert("$regex", format!("^{prefix}"));
        let mut filter = Document::new();
        filter.insert(NAMESPACE_FIELD, namespace.as_str());
        filter.insert(COLLECTION_FIELD, collection.as_str());
        filter.insert(KEY_FIELD, key_filter);
        let find_options = FindOptions::builder()
            .sort({
                let mut sort = Document::new();
                sort.insert(KEY_FIELD, 1);
                sort
            })
            .build();
        let mut cursor = store
            .find(filter, Some(find_options))
            .await
            .map_err(Self::map_error)?;
        let mut values = Vec::new();
        while cursor.advance().await.map_err(Self::map_error)? {
            let document = cursor.deserialize_current().map_err(Self::map_error)?;
            let key_text = document
                .get_str(KEY_FIELD)
                .map_err(|error| YugenDbError::driver_error("mongodb", error.to_string()))?
                .to_owned();
            let value = Self::value_from_document(document)?;
            if value.is_expired() && !options.include_expired {
                continue;
            }
            values.push((Key::try_from(key_text)?, value));
            if let Some(limit) = options.limit {
                if values.len() >= limit {
                    break;
                }
            }
        }
        Ok(values)
    }

    async fn batch(&self, batch: Batch) -> Result<BatchResult> {
        let mut set_count = 0;
        let mut delete_count = 0;
        for operation in batch.into_operations() {
            match operation {
                BatchOperation::Set {
                    namespace,
                    collection,
                    key,
                    value,
                } => {
                    self.set(
                        &namespace,
                        &collection,
                        &key,
                        value,
                        &WriteOptions::default(),
                    )
                    .await?;
                    set_count += 1;
                }
                BatchOperation::Delete {
                    namespace,
                    collection,
                    key,
                } => {
                    if self
                        .delete(&namespace, &collection, &key, &DeleteOptions::default())
                        .await?
                    {
                        delete_count += 1;
                    }
                }
            }
        }
        Ok(BatchResult::new(set_count, delete_count))
    }
}

fn regex_escape(value: &str) -> String {
    let mut escaped = String::new();
    for character in value.chars() {
        match character {
            '.' | '+' | '*' | '?' | '^' | '$' | '(' | ')' | '[' | ']' | '{' | '}' | '|' | '\\' => {
                escaped.push('\\');
                escaped.push(character);
            }
            other => escaped.push(other),
        }
    }
    escaped
}
